### ISSUE TITLE
Failing to restore net

### ISSUE THREAD

**Issue by kalle** [tensorflow/tensorflow#3926]
### Environment info
Operating System:
Mac OS X 10.11.4
Tensorflow installed from pre-built pip (no CUDA):
0.10.0rc0

### Problem
I have problems with restoring my net [(from SO)](http://stackoverflow.com/questions/39035041/trouble-restoring-checkpointed-tensorflow-net). I have created a short [test program](https://github.com/tensorflow/tensorflow/files/427240/bug_report.zip) that have the same problem as my real program.

The program train a net, run one inference with it and then checkpoints the model. Then it loads the checkpointed model and run the inference with the same data and compare the result. I expected the results to be very similar but they weren't:

    Restoring graph from /tmp/bugreport/model.ckpt-0
    Inference after training gave 2.40740537643
    Inference after restoring net gave 62579.6210938

```python
import os

import numpy as np
import tensorflow as tf
import tensorflow.contrib.layers as contrib

FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_string('tmp_dir', '/tmp/bugreport', """Temp dir""")

SIZE = 5
NUM_DATA = 10
BATCH_SIZE = 1
DEPTH = 1

def inference(input_tensor, is_training):
    "Define a stupid net"
    bn_params = {
            "is_training": is_training,
            "center": True,
            "scale": True
            }
    tensor = contrib.convolution2d(input_tensor, 8, 3,
            normalizer_fn=contrib.batch_norm,
            normalizer_params=bn_params,
            scope="conv1")
    tensor = tf.reduce_sum(tensor)
    return tensor

def training(input_data, label_data, test_data):
    """1. Train the net
    2. Do an inference pass with the trained net
    3. Checkpoint the trained net
    """
    with tf.Graph().as_default():
        input_tensor = tf.placeholder(tf.float32, [BATCH_SIZE, SIZE, SIZE, DEPT#
        label_tensor = tf.placeholder(tf.float32, [BATCH_SIZE, DEPTH], name="la#

        output = inference(input_tensor, tf.constant(True))
        loss = tf.nn.l2_loss(output - label_tensor, name="loss")
        train_op = tf.train.AdamOptimizer(0.9999).minimize(loss)
        init = tf.initialize_all_variables()
        saver = tf.train.Saver(tf.all_variables())

        with tf.Session() as sess:
            sess.run([init])
            for i in range(NUM_DATA):
                _ = sess.run(train_op, { input_tensor: input_data[i],
                                         label_tensor: label_data[i] })
            training_out = sess.run(output, { input_tensor: test_data })
            cp_path = os.path.join(FLAGS.tmp_dir, "model.ckpt")
            saver.save(sess, cp_path,
                       global_step=0, write_meta_graph=None)
    return training_out

def use_restored_net(test_data):
    """1. Load checkpointed net
    2. Do an inference pass with the trained net
    """
    with tf.Graph().as_default():
        input_tensor = tf.placeholder(tf.float32, [BATCH_SIZE, SIZE, SIZE, DEPT#

        output = inference(input_tensor, tf.constant(False))
        init = tf.initialize_all_variables()

        ckpt = tf.train.get_checkpoint_state(FLAGS.tmp_dir)
        if ckpt and ckpt.model_checkpoint_path:
            print "Restoring graph from {}".format(ckpt.model_checkpoint_path)
        else:
            raise ValueError("Could not find a checkpointed model")
        saver = tf.train.Saver(tf.all_variables())
        with tf.Session() as sess:
            sess.run([init])
            saver.restore(sess, ckpt.model_checkpoint_path)
            inference_out = sess.run(output, { input_tensor: test_data })
    return inference_out

def main(argv):
    if tf.gfile.Exists(FLAGS.tmp_dir):
        tf.gfile.DeleteRecursively(FLAGS.tmp_dir)
    tf.gfile.MakeDirs(FLAGS.tmp_dir)

    input_data = np.random.rand(NUM_DATA, BATCH_SIZE, SIZE, SIZE, DEPTH)
    label_data = np.random.rand(NUM_DATA, BATCH_SIZE, DEPTH)
    test_data = np.random.rand(BATCH_SIZE, SIZE, SIZE, DEPTH)

    training_out = training(input_data, label_data, test_data)
    inference_out = use_restored_net(test_data)
    print "Inference after training gave {}".format(training_out)
    print "Inference after restoring net gave {}".format(inference_out)


if __name__ == '__main__':
    tf.app.run()

```

**Comment by yaroslavvb** [comment]
Perhaps due to difference in models? (ie, is_training=True vs is_training=False)

**Comment by kalle** [comment]
No, same thing when setting is_training=True on the restored net.

And BTW, isn't the purpose of the `is_training` argument to batch normalization to "integrate" the learned mean and std into the layer during interference?

**Comment by kalle** [comment]
@yaroslavvb I was too quick when I said it was the same result (I must have forgotten to save the file of something). When I set `is_training=True` I get the same result when doing the inference with the restored net.

But I still don't understand the purpose of the `is_training` parameter since setting it to `False` essentially throws away the learned mean and variance, replace them with 0.0 and 1.0 and then do the normalization. If this is the intended behavior of `is_training` (but the doc string indicates that it isn't) perhaps it should be called `use_learned_data` instead.

**Comment by yaroslavvb** [comment]
During training there are two sources of mean and variable -- statistics over mini-batch, and global statistics computed  using geometric averaging. The idea is that while training you use mini-batch statistics to normalize, and just save global statistics for later. During inference you may not have mini-batches, so you global statistics instead. The name `is_training` is also used for dropout layers which has different behavior depending on whether you are training

**Comment by kalle** [comment]
Yaroslav , thank your for taking the time to answer my questions on a weekend! :-)

OK, I have made another [test program](https://github.com/tensorflow/tensorflow/files/428839/bug_report2.zip) that only uses a batch normalizer and I train beta and gamma to give the the distribution I want.

I send in noise with distribution Norm(m=2,v=10) and then learn the parameters beta and gamma so the output from the batch normalization is distributed as Norm(m=7,v=5). Then I run an inference pass with test data directly after training I get the expected mean and variance (7 resp. 5).
But when I restore the net and send in the same test data I get some weird values that don't make any sense at all. Even if the input data is generated with Norm(m=2,v=10) which BatchNorm should have learnt by now. The moving mean and variance don't appear to be updated and the only reason that BN work when I run it in training mode is that it uses the batch mean/variance.

```
test data:
    mean = [ 1.99910051  2.00070378]
    variance = [  9.99379123  10.02349832]
After training:
    calculated mean=[ 7.01989126  7.01992226]
    calculated variance=[ 4.9753685   4.97816467]
    beta=[ 7.01995707  7.0199585 ]
    gamma=[ 2.23007703  2.23052549]
    moving mean=[ 0.  0.]
    moving variance=[ 1.  1.]
After restore:
    calculated mean=[ 11.47598934  11.4803896 ]
    calculated variance=[ 49.58828735  49.75190735]
    beta=[ 7.01995707  7.0199585 ]
    gamma=[ 2.23007703  2.23052549]
    moving mean=[ 0.  0.]
    moving variance=[ 1.  1.]
```

Test program:

```python
import os

import numpy as np
import tensorflow as tf
import tensorflow.contrib.layers as contrib

FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_string('tmp_dir', '/tmp/bugreport', """Temp dir""")

SIZE = 500
NUM_ITERS = 100
BATCH_SIZE = 5
DEPTH = 2
MEAN = 2
VARIANCE = 10

def _randn(*shape):
    "Create a tensor with values drawn from a gaussion distribution (MEAN, VARIANCE)"
    return np.sqrt(VARIANCE) * np.random.randn(*shape) + MEAN


def _get_bn_vars(sess):
    "Return values of internal BatchNorm variables"
    beta = None
    gamma = None
    moving_mean = None
    moving_variance = None
    for v in tf.all_variables():
        if v.name == "BatchNorm/beta:0":
            beta = sess.run(v)
        elif v.name == "BatchNorm/gamma:0":
            gamma = sess.run(v)
        elif v.name == "BatchNorm/moving_mean:0":
            moving_mean = sess.run(v)
        elif v.name == "BatchNorm/moving_variance:0":
            moving_variance = sess.run(v)
    return beta, gamma, moving_mean, moving_variance


def inference(input_tensor, is_training):
    "Define a stupid net"
    bn_out = contrib.batch_norm(input_tensor,
            center=True,
            scale=True,
            is_training=is_training)
    return bn_out


def training(test_data):
    """1. Train the net
    2. Do an inference pass with the trained net
    3. Checkpoint the trained net
    """
    with tf.Graph().as_default():
        tf.set_random_seed(123)
        input_tensor = tf.placeholder(tf.float32, [BATCH_SIZE, SIZE, SIZE, DEPTH], name="input")

        bn_out = inference(input_tensor, tf.constant(True))
        mean, variance = tf.nn.moments(bn_out, [0, 1, 2])
        loss = tf.nn.l2_loss(mean-7) + tf.nn.l2_loss(variance - 5)
        train_op = tf.train.AdamOptimizer(0.9999).minimize(loss)
        init = tf.initialize_all_variables()
        saver = tf.train.Saver(tf.all_variables())

        with tf.Session() as sess:
            sess.run([init])
            for _ in range(NUM_ITERS):
                input_data = _randn(BATCH_SIZE, SIZE, SIZE, DEPTH)
                _ = sess.run(train_op, { input_tensor: input_data } )
            mean_val, variance_val = sess.run([mean, variance], { input_tensor: test_data })
            cp_path = os.path.join(FLAGS.tmp_dir, "model.ckpt")
            saver.save(sess, cp_path,
                       global_step=0, write_meta_graph=None)
            bn_vars = _get_bn_vars(sess)
    return mean_val, variance_val, bn_vars


def use_restored_net(test_data):
    """1. Load checkpointed net
    2. Do an inference pass with the trained net
    """
    with tf.Graph().as_default():
        input_tensor = tf.placeholder(tf.float32, [BATCH_SIZE, SIZE, SIZE, DEPTH], name="input")

        bn_out = inference(input_tensor, tf.constant(False))
        mean, variance = tf.nn.moments(bn_out, [0, 1, 2])
        init = tf.initialize_all_variables()

        ckpt = tf.train.get_checkpoint_state(FLAGS.tmp_dir)
        if not(ckpt and ckpt.model_checkpoint_path):
            raise ValueError("Could not find a checkpointed model")
        saver = tf.train.Saver(tf.all_variables())
        with tf.Session() as sess:
            sess.run([init])
            saver.restore(sess, ckpt.model_checkpoint_path)
            bn_vars = _get_bn_vars(sess)
            mean_val, variance_val = sess.run([mean, variance], { input_tensor: test_data })
    return mean_val, variance_val, bn_vars


def main(argv):
    if tf.gfile.Exists(FLAGS.tmp_dir):
        tf.gfile.DeleteRecursively(FLAGS.tmp_dir)
    tf.gfile.MakeDirs(FLAGS.tmp_dir)

    np.random.seed(123)
    test_data = _randn(BATCH_SIZE, SIZE, SIZE, DEPTH)
    print "test data:"
    print "    mean = {}".format(np.mean(test_data, axis=(0, 1, 2)))
    print "    variance = {}".format(np.var(test_data, axis=(0, 1, 2)))

    test_mean, test_variance, train_bn = training(test_data)
    print "After training:"
    print "    calculated mean={}".format(test_mean)
    print "    calculated variance={}".format(test_variance)
    print "    beta={}".format(train_bn[0])
    print "    gamma={}".format(train_bn[1])
    print "    moving mean={}".format(train_bn[2])
    print "    moving variance={}".format(train_bn[3])
    restored_mean, restored_variance, restored_bn = use_restored_net(test_data)
    print "After restore:"
    print "    calculated mean={}".format(restored_mean)
    print "    calculated variance={}".format(restored_variance)
    print "    beta={}".format(restored_bn[0])
    print "    gamma={}".format(restored_bn[1])
    print "    moving mean={}".format(restored_bn[2])
    print "    moving variance={}".format(restored_bn[3])


if __name__ == '__main__':
    tf.app.run()

```

**Comment by yaroslavvb** [comment]
So you are saying that moving_mean and moving_variance don't get updated when training? You can see the code in [layers.py](https://github.com/tensorflow/tensorflow/blob/008bcaea38815f46804fc3f56492f4dd93837a56/tensorflow/contrib/layers/python/layers/layers.py#L229). There's a node there "update_moving_mean" and as far as I can see, it should be triggered whenever you read output of batchnorm. You could add some passthrough Print nodes to that code (ie, do `import inspect;
inspect.getsourcefile(tf.contrib.layers)` and modify `layers.py` in that directory) to double-check that those nodes get called

**Comment by kalle** [comment]
Yes, I guess that is was I'm saying.

I added the print outs and nothing got printed out. But when I changed this lines in `optimizer.Optimizer.apply_gradients()`

```python
update_ops = []
```

to

```python
update_ops = list(ops.get_collection(ops.GraphKeys.UPDATE_OPS))
```

the batch normalizer did learn the distribution of its input (and got correctly restored)! I don't know if this is the correct place to fetch the `UPDATE_OPS` or if there needs to be more checks or whatever.
